spec/e2e/association_proxy_spec.rb

```diff
@@ -107,6 +115,26 @@
     end
   end
 
+  it 'Queries limited times in depth two loops with deep with_associations' do
+    Student.create.lessons << science
+    Student.create.lessons << science
+    expect_queries(1) do
+      science.students.with_associations(lessons: :exams_given).each do |student|
+        student.lessons.each do |lesson|
+          lesson.exams_given.to_a
+        end
+      end
+    end
+  end
+
+  it 'Queries only one time when there are some empty associations' do
+    Student.create.lessons << science
+    Student.create.lessons += [science, Lesson.create]
+    expect_queries(1) do
+      science.students.with_associations(lessons: :exams_given).flat_map(&:lessons).flat_map(&:exams_given)
+    end
+  end
```

Comment:
Currently, on master `each_rel` called after `with_association` does not work properly. It wrongfully behaves like `each` passing a node to the block. Here I was actually not attempting to support `each_rel`, but maybe we should fix that for the sake of consistency.